# Operations Research: Models & Applications

## Simple LP Formulations

### Problem 1 - Product Mix

- Objective: Maximise revenue from items made. £700 for a desk and £900 for a table.
- Decision variables - x1 (number of desks) & x2 (number of tables)
- Constraint 1: 3600 wood units. Each desk needs 3 units and a table 5.
- Constraint 2: 1600 labour hours available. Desk needs 1hr and Tables 2.
- Constraint 3: 800 machine hours available. 50 mins for a desk and 20 for a table

System of equations:

```
  max  700x1 + 900x2  
  s.t.   3x1 +   5x2 <= 3600  
          x1 +   2x2 <= 1600  
        50x1 +  20x2 <= 48000  
          x1         >= 0  
                  x2 >= 0  
```

Transformed system of equations (to have only <= constraints and a minimisation objective): 

```
  min  -700x1 - 900x2  
  s.t. +  3x1 +   5x2 <= +3600  
       +   x1 +   2x2 <= +1600  
       + 50x1 +  20x2 <= +48000  
          x1          >=  0  
                   x2 >=  0  
```

Test one of the possible solvers.

In [4]:
! pulptest

ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss	 Test that logic put in place for deprecation handling of indexs works
.	 Testing 'indexs' param continues to work for LpVariable.dicts
	 Testing 'indexs' param continues to work for LpVariable.matrix
.	 Testing 'indices' argument works in LpVariable.dicts
	 Testing 'indices' param continues to work for LpVariable.matrix
.	 Testing invalid status
.	 Testing continuous LP solution - export dict
.	 Testing export dict for LP
.	 Testing e

In [5]:
from scipy.optimize import linprog

In [17]:
obj = [-700, -900]
lhs_ineq = [
    [3, 5],
    [1, 2],
    [50, 20]
]
rhs_ineq = [
    3600,
    1600,
    48000
]
bnd = [
    (0, float("inf")),  # Bounds of x1
    (0, float("inf"))]  # Bounds of x2


In [20]:
opt = linprog(
    c=obj, A_ub=lhs_ineq, b_ub=rhs_ineq,
    # A_eq=lhs_eq, b_eq=rhs_eq, 
    bounds=bnd,
    method="revised simplex")

opt

     con: array([], dtype=float64)
     fun: -789473.6842105263
 message: 'Optimization terminated successfully.'
     nit: 2
   slack: array([  0.        , 336.84210526,   0.        ])
  status: 0
 success: True
       x: array([884.21052632, 189.47368421])

From the output, of `opt.x` for x1 = 884.2 and x2 = 189.4, we have an optimum of £789,473. These figures could be rounded down and used as a starting point.